### ISSUE TITLE
Allow arbitrary types for a subset of atomic operations

### ISSUE THREAD

**Issue by jordanisaacs** [ziglang/zig#19576]
Atomic operations `@store`, `@load`, `@cmpxchgWeak`, `@cmpxchStrong`, and `@atomicRmw(.Xchg)` are currently only allowed for: `bool, integer, float, enum, or pointer type`. I am proposing allowing arbitrary types, with the requirement that the size corresponds to an atomic operation and alignment is correct.

This is sound because retrieval operations don't change the data representation. And storing data will be as valid as its input which is provided by the typesystem.

The use-case for this revolves around the standard library letting you wrap types to make them "atomic" using `std.atomic.Value(type)`. However, you can only wrap primitive types because of the limitations of the atomic builtins. You can't provide pointer wrappers or packed structs. The following gives you a compile error, but with this proposal it will compile.

```zig
const std = @import("std");

const PtrWrapper = packed struct(usize) {
    ptr: *const usize,

    fn is_10(self: *@This()) bool {
        return *self.ptr == 10;
    }
};

const PackedStruct = packed struct(u64) {
    flag1: bool,
    flag2: u63,
};

test "arbitrary atomics" {
    const AtomicPtrWrapper = std.atomic.Value(PtrWrapper);
    const AtomicPackedStruct = std.atomic.Value(PackedStruct);

    const x: usize = 10;
    const y: usize = 20;
    var atomicPtr = AtomicPtrWrapper.init(.{ .ptr = &x });
    try std.testing.expect(atomicPtr.swap(.{ .ptr = &y }, .acq_rel).is_10());

    const flag = AtomicPackedStruct.init(.{ .flag1 = true, .flag2 = 100 });
    try std.testing.expectEqual(true, flag.load(.acq_rel).flag1);
}
```

The equivalent today requires creating separate types ala `PackedStruct` & `AtomicPackedStruct` and re-exposing the atomic methods. This is seen in the following gist since I can't just do `std.atomic.Value(InvalidPtr)`: https://gist.github.com/jordanisaacs/a121a081a3623270e387dae32b08486e.

**Comment by kprotty** [comment]
Why not use `atomic.Value(u64/usize)` then `@bitCast/ptrFromInt` into the packed struct or pointer? This is what [zap](https://github.com/mitchellh/libxev/blob/17a910a2d01698ae1a3aa1a39da5ea93f36fd2f3/src/ThreadPool.zig#L195-L197) does for example.

**Comment by Rexicon226** [comment]
> You can't provide pointer wrappers or packed structs. The following gives you a compile error, but with this proposal it will compile.

This was implemented in #20590.

Can this be closed?

**Comment by jordanisaacs** [comment]
Yep, it now works! Here is an updated gist with it working 
https://gist.github.com/jordanisaacs/4472c187f3a5903279c53e56b527d970
